File: exercism/clojure/phone-number/src/phone_number.clj
(ns phone-number
  (:require [clojure.string :as s]))

(defn drop-country-code [number]
  (if (= 11 (count number))
    (s/replace number #"^1" "")
    number))

(defn number [num-string]
  (let [number (->> num-string
                (re-seq #"\d+")
                (apply str)
                (drop-country-code))]
    (if (= (count number) 10)
      number
      "0000000000")))

(defn parts [number]
  (->> number
    (re-matches #"(\d{3})(\d{3})(\d{4})")
    (rest)))

(defn area-code [num-string]
  (->> num-string
    (number)
    (parts)
    (first)))

(defn pretty-print
  ([num-string]
   (->> num-string
     (number)
     (parts)
     (apply pretty-print)))
  ([area-code exchange-code subscriber]
   (format "(%s) %s-%s" area-code exchange-code subscriber)))

File: exercism/javascript/phone-number/phone-number.js
const areaCode = nr => nr[0];
const exchangeCode = nr => nr[3];

export default class PhoneNumber {
  constructor(nr) {
    this.nr = nr;
  }

  number() {
    const nr = this.nr.replace(/^(\+1|1)|(\D)/g, '');

    switch (true) {
      case nr.length !== 10:
        return null;
      case areaCode(nr) === '0' || areaCode(nr) === '1':
        return null;
      case exchangeCode(nr) === '0' || exchangeCode(nr) === '1':
        return null;
      default:
        return nr;
    }
  }
}
